File: bindings/cpp/src/resp_conn.cpp
#include "resp_conn.hpp"

#include <netdb.h>
#include <netinet/in.h>
#include <netinet/tcp.h>
#include <sys/socket.h>
#include <sys/time.h>
#include <unistd.h>

#include <cerrno>
#include <cstring>

#include "kevy/errors.hpp"
#include "kevy/resp.hpp"

namespace kevy {
namespace detail {

std::unique_ptr<RespConn> RespConn::dial(const std::string& host, uint16_t port) {
  struct addrinfo hints{};
  hints.ai_family = AF_UNSPEC;
  hints.ai_socktype = SOCK_STREAM;
  struct addrinfo* res = nullptr;
  std::string port_s = std::to_string(port);
  if (getaddrinfo(host.c_str(), port_s.c_str(), &hints, &res) != 0 || res == nullptr) {
    throw IoError("kevy: getaddrinfo failed for " + host);
  }
  int fd = -1;
  for (struct addrinfo* p = res; p != nullptr; p = p->ai_next) {
    fd = ::socket(p->ai_family, p->ai_socktype, p->ai_protocol);
    if (fd < 0) continue;
    if (::connect(fd, p->ai_addr, p->ai_addrlen) == 0) break;
    ::close(fd);
    fd = -1;
  }
  freeaddrinfo(res);
  if (fd < 0) throw IoError("kevy: connect failed for " + host + ":" + port_s);
  int one = 1;
  ::setsockopt(fd, IPPROTO_TCP, TCP_NODELAY, &one, sizeof(one));
  return std::unique_ptr<RespConn>(new RespConn(fd));
}

void RespConn::close() {
  if (fd_ >= 0) {
    ::close(fd_);
    fd_ = -1;
  }
}

void RespConn::write_all(const char* data, size_t len) {
  size_t sent = 0;
  while (sent < len) {
    ssize_t n = ::send(fd_, data + sent, len - sent, 0);
    if (n > 0) {
      sent += static_cast<size_t>(n);
      continue;
    }
    if (n < 0 && errno == EINTR) continue;
    close();
    throw IoError(std::string("kevy: send failed: ") + std::strerror(errno));
  }
}

Reply RespConn::read_one() {
  char chunk[8192];
  for (;;) {
    Reply r;
    size_t used = 0;
    try {
      used = resp::parse_reply(reinterpret_cast<const uint8_t*>(rbuf_.data()), rbuf_.size(), r);
    } catch (const KevyError&) {
      throw ProtocolError("malformed reply");
    }
    if (used > 0) {
      rbuf_.erase(0, used);
      return r;
    }
    ssize_t n = ::recv(fd_, chunk, sizeof(chunk), 0);
    if (n > 0) {
      rbuf_.append(chunk, static_cast<size_t>(n));
      continue;
    }
    if (n == 0) {
      close();
      throw ClosedError();  // clean server close mid-read
    }
    if (errno == EINTR) continue;
    if (errno == EAGAIN || errno == EWOULDBLOCK) throw TimedOutError();
    close();
    throw IoError(std::string("kevy: recv failed: ") + std::strerror(errno));
  }
}

Reply RespConn::request(const std::vector<std::string>& argv) {
  if (fd_ < 0) throw ClosedError();
  std::string wire;
  resp::encode_command(wire, argv);
  write_all(wire.data(), wire.size());
  return read_one();
}

std::vector<Reply> RespConn::pipeline_raw(const std::string& wire, size_t n) {
  if (fd_ < 0) throw ClosedError();
  write_all(wire.data(), wire.size());
  std::vector<Reply> out;
  out.reserve(n);
  for (size_t i = 0; i < n; i++) out.push_back(read_one());
  return out;
}

void RespConn::write(const std::vector<std::string>& argv) {
  if (fd_ < 0) throw ClosedError();
  std::string wire;
  resp::encode_command(wire, argv);
  write_all(wire.data(), wire.size());
}

Reply RespConn::read_reply() {
  if (fd_ < 0) throw ClosedError();
  return read_one();
}

void RespConn::set_read_timeout(std::optional<uint64_t> ms) {
  if (fd_ < 0) throw ClosedError();
  struct timeval tv{};
  if (ms.has_value()) {
    tv.tv_sec = static_cast<time_t>(*ms / 1000);
    tv.tv_usec = static_cast<suseconds_t>((*ms % 1000) * 1000);
  }
  ::setsockopt(fd_, SOL_SOCKET, SO_RCVTIMEO, &tv, sizeof(tv));
}

void RespConn::select_db(uint32_t db) {
  Reply r = request({"SELECT", std::to_string(db)});
  if (r.kind == ReplyKind::Error) {
    throw ProtocolError("SELECT " + std::to_string(db) + " rejected: " + r.str());
  }
}

}  // namespace detail
}  // namespace kevy

File: bindings/go/resp_conn.go
package kevy

import (
	"context"
	"errors"
	"io"
	"net"
	"strconv"
	"time"
)

// respConn is a blocking RESP2/RESP3 connection over TCP — the native
// remote transport (contract §1.1). One request at a time; a read buffer
// reassembles multi-segment replies across reads. Not safe for concurrent
// use by multiple goroutines.
type respConn struct {
	conn    net.Conn
	rbuf    []byte
	wbuf    []byte
	hostStr string
}

// dialResp opens a TCP RESP connection with TCP_NODELAY.
func dialResp(host string, port uint16) (*respConn, error) {
	addr := net.JoinHostPort(host, strconv.Itoa(int(port)))
	conn, err := net.Dial("tcp", addr)
	if err != nil {
		return nil, errIo(err)
	}
	if tc, ok := conn.(*net.TCPConn); ok {
		_ = tc.SetNoDelay(true)
	}
	return &respConn{
		conn:    conn,
		rbuf:    make([]byte, 0, 8192),
		wbuf:    make([]byte, 0, 1024),
		hostStr: addr,
	}, nil
}

func (c *respConn) close() {
	if c.conn != nil {
		_ = c.conn.Close()
		c.conn = nil
	}
}

// encodeCommand writes argv as a RESP multibulk request.
func encodeCommand(out []byte, argv [][]byte) []byte {
	out = append(out, '*')
	out = strconv.AppendInt(out, int64(len(argv)), 10)
	out = append(out, '\r', '\n')
	for _, a := range argv {
		out = append(out, '$')
		out = strconv.AppendInt(out, int64(len(a)), 10)
		out = append(out, '\r', '\n')
		out = append(out, a...)
		out = append(out, '\r', '\n')
	}
	return out
}

// request sends one command and blocks until exactly one reply is parsed.
// ctx bounds the call: a deadline arms the socket, cancellation interrupts
// the in-flight read. A clean server close mid-read surfaces as Closed.
func (c *respConn) request(ctx context.Context, argv [][]byte) (Reply, error) {
	if c.conn == nil {
		return Reply{}, errClosed()
	}
	c.wbuf = encodeCommand(c.wbuf[:0], argv)
	replies, err := c.roundtrip(ctx, c.wbuf, 1)
	if err != nil {
		return Reply{}, err
	}
	return replies[0], nil
}

// pipelineRaw sends a pre-encoded batch as one write and reads n replies.
func (c *respConn) pipelineRaw(ctx context.Context, raw []byte, n int) ([]Reply, error) {
	if c.conn == nil {
		return nil, errClosed()
	}
	return c.roundtrip(ctx, raw, n)
}

func (c *respConn) roundtrip(ctx context.Context, wire []byte, n int) ([]Reply, error) {
	stop := c.armContext(ctx)
	defer stop()
	if err := ctxErr(ctx); err != nil {
		return nil, err
	}
	if _, err := c.conn.Write(wire); err != nil {
		return nil, c.mapIoErr(ctx, err)
	}
	out := make([]Reply, 0, n)
	for len(out) < n {
		r, err := c.readOne(ctx)
		if err != nil {
			return nil, err
		}
		out = append(out, r)
	}
	return out, nil
}

func (c *respConn) readOne(ctx context.Context) (Reply, error) {
	chunk := make([]byte, 8192)
	for {
		r, used, perr := parseReply(c.rbuf)
		if perr != nil {
			return Reply{}, errProtocol("malformed reply")
		}
		if used > 0 {
			c.rbuf = c.rbuf[used:]
			return r, nil
		}
		n, err := c.conn.Read(chunk)
		if n > 0 {
			c.rbuf = append(c.rbuf, chunk[:n]...)
			continue
		}
		if err != nil {
			return Reply{}, c.mapIoErr(ctx, err)
		}
	}
}

// armContext wires ctx cancellation/deadline onto the socket. On a
// deadline it sets the socket deadline; on cancellation a watcher
// interrupts the blocked read by pushing the deadline into the past
// (leaving the connection reusable). Returns a stop func.
func (c *respConn) armContext(ctx context.Context) func() {
	if ctx == nil {
		return func() {}
	}
	if dl, ok := ctx.Deadline(); ok {
		_ = c.conn.SetDeadline(dl)
	}
	done := ctx.Done()
	if done == nil {
		return func() { _ = c.conn.SetDeadline(time.Time{}) }
	}
	stopped := make(chan struct{})
	go func() {
		select {
		case <-done:
			_ = c.conn.SetReadDeadline(time.Now())
		case <-stopped:
		}
	}()
	return func() {
		close(stopped)
		_ = c.conn.SetDeadline(time.Time{})
	}
}

func (c *respConn) mapIoErr(ctx context.Context, err error) error {
	if errors.Is(err, io.EOF) {
		c.close()
		return errClosed()
	}
	var ne net.Error
	if errors.As(err, &ne) && ne.Timeout() {
		if ctx != nil && ctx.Err() != nil {
			if errors.Is(ctx.Err(), context.Canceled) {
				return &KevyError{Kind: KindClosed, Msg: "context canceled"}
			}
			return &KevyError{Kind: KindTimedOut}
		}
		return &KevyError{Kind: KindTimedOut}
	}
	c.close()
	return errIo(err)
}

func ctxErr(ctx context.Context) error {
	if ctx == nil {
		return nil
	}
	if err := ctx.Err(); err != nil {
		if errors.Is(err, context.DeadlineExceeded) {
			return &KevyError{Kind: KindTimedOut}
		}
		return &KevyError{Kind: KindClosed, Msg: "context canceled"}
	}
	return nil
}

// write sends argv as one command without reading a reply — the
// decoupled send side used by the pub/sub consumer, which then drains
// frames with readReply.
func (c *respConn) write(argv [][]byte) error {
	if c.conn == nil {
		return errClosed()
	}
	c.wbuf = encodeCommand(c.wbuf[:0], argv)
	if _, err := c.conn.Write(c.wbuf); err != nil {
		return c.mapIoErr(context.Background(), err)
	}
	return nil
}

// readReply blocks for the next reply frame, honouring ctx and any socket
// read deadline set via setReadDeadline. Used by the pub/sub consumer.
func (c *respConn) readReply(ctx context.Context) (Reply, error) {
	if c.conn == nil {
		return Reply{}, errClosed()
	}
	stop := c.armContext(ctx)
	defer stop()
	return c.readOne(ctx)
}

// setReadDeadline arms a bounded blocking read (or clears it with zero).
func (c *respConn) setReadDeadline(d time.Time) error {
	if c.conn == nil {
		return errClosed()
	}
	return c.conn.SetReadDeadline(d)
}

// selectDB issues a SELECT n round-trip; a -ERR reply fails the connect.
func (c *respConn) selectDB(db uint32) error {
	r, err := c.request(context.Background(), [][]byte{[]byte("SELECT"), []byte(strconv.FormatUint(uint64(db), 10))})
	if err != nil {
		return err
	}
	if r.Kind == KindError {
		return errProtocol("SELECT %d rejected: %s", db, r.Str())
	}
	return nil
}
